### PR TITLE
[Patterns]: Support the `blockTypes` prop for patterns fetched from Pattern Directory

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
This PR adds support for the `blockTypes` prop 
The changes here are from this PR: https://github.com/WordPress/gutenberg/pull/32113 and is just here to move things faster in order to land this for 6.2.

props @ryelle 🙇

Without this change, the patterns fetched from Pattern Directory wouldn't use this prop when they were registered. This become more obvious with this issue for [adding more Query Loop patterns](https://github.com/WordPress/gutenberg/issues/44140) and it affects the `header, footer` patterns that are now fetched from Pattern Directory here: https://github.com/WordPress/gutenberg/issues/46017

Also noting that the added patterns from https://github.com/WordPress/gutenberg/issues/46017, seem to still don't have the `blockTypes` prop in PD and they would need to be updated to include this.

#### Before

https://user-images.githubusercontent.com/16275880/216287451-cbadec15-3a7e-4461-bb8e-d3e8ea7f9aab.mov

#### After

https://user-images.githubusercontent.com/16275880/216287480-253f724b-4605-4387-9796-8767cd0fa080.mov

## Testing Instructions
1. Insert a Heading block
2. Open the block switcher menu
3. Observe that there is a pattern suggested, coming from Pattern Directory

### CODE REVIEW

**Review by carolinan — COMMENTED** [review]
The feature with the transform works well.
I am unable to review the test.

In gutenberg_load_remote_block_patterns the viewport_width setting seems unrelated to enabling block types -but I don't see a reason to remove it if it speeds things up.

**Comment by ntsekouras** [comment]
>In gutenberg_load_remote_block_patterns the viewport_width setting seems unrelated to enabling block types -but I don't see a reason to remove it if it speeds things up.

Yes, it is, but it was caught from @ryelle and I think it's land this too.

File: lib/compat/wordpress-6.2/block-patterns.php [hunk 1]
```diff
@@ -310,3 +310,60 @@ function gutenberg_register_theme_block_patterns() {
...
+		}
+		$patterns = $response->get_data();
+
+		foreach ( $patterns as $settings ) {
+			if ( isset( $settings['block_types'] ) ) {
+				$settings['blockTypes'] = $settings['block_types'];
+				unset( $settings['block_types'] );
+			}
+
+			if ( isset( $settings['viewport_width'] ) ) {
```
Comment (ntsekouras) [inline lib/compat/wordpress-6.2/block-patterns.php]: This setting is not directly related to the main `block_types` change, but it's okay fix this too.

File: lib/compat/wordpress-6.2/block-patterns.php [hunk 2]
```diff
@@ -310,3 +310,60 @@ function gutenberg_register_theme_block_patterns() {
...
+		$request         = new WP_REST_Request( 'GET', '/wp/v2/pattern-directory/patterns' );
+		$core_keyword_id = 11; // 11 is the ID for "core".
+		$request->set_param( 'keyword', $core_keyword_id );
+		$response = rest_do_request( $request );
+		if ( $response->is_error() ) {
+			return;
+		}
+		$patterns = $response->get_data();
+
+		foreach ( $patterns as $settings ) {
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.2/block-patterns.php]: Nit: Can we change `$settings` to a `$pattern`? So it matches other `_load_remote_*` functions. See https://github.com/WordPress/wordpress-develop/blob/6.1/src/wp-includes/block-patterns.php#L118-L159.

File: phpunit/class-wp-rest-pattern-directory-controller-test.php [hunk 3]
```diff
@@ -212,23 +212,74 @@ function ( $preempt, $args, $url ) {
...
+		$this->assertPatternMatchesSchema( $prepared_pattern );
+		$this->assertArrayNotHasKey( 'extra_field', $prepared_pattern );
+	}
+
+	/**
+	 * Asserts that the pattern matches the expected response schema.
+	 *
+	 * @param WP_REST_Response[] $pattern An individual pattern from the REST API response.
+	 */
+	public function assertPatternMatchesSchema( $pattern ) {
```
Comment (Mamaduka) [inline phpunit/class-wp-rest-pattern-directory-controller-test.php:237]: I think we already have this in core tests 👍

File: lib/compat/wordpress-6.2/rest-api.php [hunk 4]
```diff
@@ -108,3 +108,43 @@ function gutenberg_modify_rest_sidebars_response( $response ) {
...
+				'description' => __( 'The block types which can use this pattern.', 'gutenberg' ),
+				'type'        => 'array',
+				'uniqueItems' => true,
+				'items'       => array( 'type' => 'string' ),
+				'context'     => array( 'view', 'embed' ),
+			),
+		)
+	);
+}
+add_filter( 'rest_api_init', 'add_block_pattern_block_types_schema' );
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.2/rest-api.php:133]: We already extend `WP_REST_Pattern_Directory_Controller` in 6.2 compat dir. Maybe we should change the schema and response there.

What do you think?

**Review by Mamaduka — APPROVED** [review]
I left a few minor notes, but the fix works as expected. PHPUnit tests are passing locally as well ✅

File: lib/compat/wordpress-6.2/block-patterns.php (hunk 2 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.2/block-patterns.php]: Your comment made me realize I missed to update the other remote calls. I'll push an update shortly.

Comment (ntsekouras) [reply lib/compat/wordpress-6.2/block-patterns.php]: Updated here: https://github.com/WordPress/gutenberg/pull/47677/commits/06f6b167c7fc7589848673c210d21cc6a9036791

File: phpunit/class-wp-rest-pattern-directory-controller-test.php (hunk 3 above)
Comment (ntsekouras) [reply phpunit/class-wp-rest-pattern-directory-controller-test.php:237]: Yes, but since we are updating the schema, I think it's good to have them here.. This will not be copied to core during the back port.

File: lib/compat/wordpress-6.2/rest-api.php (hunk 4 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.2/rest-api.php:133]: When we can update something with filters, I think it's preferable as we don't need to copy the whole function.

Comment (Mamaduka) [reply lib/compat/wordpress-6.2/rest-api.php:133]: I don't have any strong preference here. Having all contextual changes in a single place makes backports easier.

File: lib/compat/wordpress-6.2/block-patterns.php [hunk 5]
```diff
@@ -310,3 +310,150 @@ function gutenberg_register_theme_block_patterns() {
...
+	$response = rest_do_request( $request );
+	if ( $response->is_error() ) {
+		return;
+	}
+	$patterns = $response->get_data();
+
+	foreach ( $patterns as $pattern ) {
+		$normalized_pattern = gutenberg_normalize_remote_pattern( $pattern );
+		$pattern_name       = sanitize_title( $normalized_pattern['title'] );
+		$registry           = WP_Block_Patterns_Registry::get_instance();
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.2/block-patterns.php]: I had this somewhere on my to-do list, so let's do it here 😄

We should move this like outside the loop; no need to get the same registry instance on each iteration.

Comment (ntsekouras) [reply lib/compat/wordpress-6.2/block-patterns.php]: Good call! I'll update

**Comment by Mamaduka** [comment]
Thank you, Nik. Updates look good 🚢

**Comment by Mamaduka** [comment]
I just cherry-picked this PR to the release/15.1 branch to get it included in the next release: f0c0581bb0

File: lib/compat/wordpress-6.2/block-patterns.php [hunk 6]
```diff
@@ -310,3 +310,149 @@ function gutenberg_register_theme_block_patterns() {
...
+	/** This filter is documented in wp-includes/block-patterns.php */
+	if ( ! apply_filters( 'should_load_remote_block_patterns', true ) ) {
+		return;
+	}
+
+	if ( ! wp_theme_has_theme_json() ) {
+		return;
+	}
+
+	$pattern_settings = WP_Theme_JSON_Resolver::get_theme_data()->get_patterns();
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/block-patterns.php:436]: @ntsekouras @Mamaduka @carolinan We should use the class bundled with Gutenberg and not the one provided by core. Prepared a fix for this at https://github.com/WordPress/gutenberg/pull/49197

Comment (ntsekouras) [reply lib/compat/wordpress-6.2/block-patterns.php:436]: Thank you @oandregal !
